### PR TITLE
fix(renderer): clean out edge cases of cursor.position.restore

### LINKED ISSUE #1310: BUG: jump to selected item when opening folder
### Did you check docs and existing issues?

- [X] I have read all the docs.
- [X] I have searched the existing issues.
- [X] I have searched the existing discussions.

### Neovim Version (nvim -v)

NVIM v0.9.5 Build type: Release LuaJIT 2.1.1693350652

### Operating System / Version

macos 14.2.1 

### Describe the Bug

some times when opening folder, cursor will jump to a random item rather than stay at the foler.
some times it jump to opened file ,sometimes it jump to item in the folder. i couldn't get exact clue.

### Screenshots, Traceback

[replicate video](https://1drv.ms/v/s!Aul7Mp-RSA34gfxgj6FKIE1ps0TmPg?e=IeXjS9)

### Steps to Reproduce

1. open the file with neotree
2. focus on neotree
3. selecte a folder , press l to open
4. the cursor jump to opened item , not stay at the folder

### Expected Behavior

the cursor should stay at the folder item.

### Your Configuration

```Lua
i am using AstroNvim 3.41.2 with my personal config
{
    "nvim-neo-tree/neo-tree.nvim",
    opts = {
      filesystem = {
        filtered_items = {
          visible = true,
          show_hidden_count = true,
          hide_dotfiles = false,
          hide_gitignored = true,
          hide_by_name = {
            -- '.git',
            -- '.DS_Store',
            -- 'thumbs.db',
          },
          never_show = {},
        },
      },
      window = {
        position = "right",
      },
    },
  }
```

### PR DESCRIPTION
**Pull request by pysan3**
We need to be more careful when to save and restore the cursor position as we cannot overwrite the position anymore with https://github.com/nvim-neo-tree/neo-tree.nvim/pull/1288.

closes #1310

Please read the LONG discussion from https://github.com/nvim-neo-tree/neo-tree.nvim/pull/1354#issuecomment-1946307225 as it describes the issue and the reason thoroughly.

I will keep this as draft until I get some people to test this code as I might have missed some more edge cases.

Would you mind testing this code? @georgeguimaraes @mehalter @XavierChanth

```lua
{ 
  "pysan3/neo-tree.nvim",
  branch = "update-curpos-before-expand",
  version = false,
```

### CODE REVIEW

**Comment by mehalter** [comment]
I did some testing and this does seem to fix the linked issue! ❤️

**Comment by XavierChanth** [comment]
Working wonderfully @pysan3 ! Thanks so much

**Comment by pysan3** [comment]
Thanks @mehalter @XavierChanth !

Ready to roll @cseickel

File: lua/neo-tree/ui/renderer.lua [hunk 1]
```diff
@@ -522,8 +522,10 @@ M.focus_node = function(state, id, do_not_focus_window, relative_movement, botto
     end
     local success, err = pcall(vim.api.nvim_win_set_cursor, state.winid, { linenr, col })
 
-    -- now ensure that the window is scrolled correctly
     if success then
+      -- forget about cursor position as it is overwritten
+      M.position.clear(state)
```
Comment (cseickel) [inline lua/neo-tree/ui/renderer.lua:527]: Couldn't we have just set 
```suggestion
      M.position.is.restorable = false
```
 at this location and accomplish the same thing without adding the clear method? Are we duplicating functionality by have both?

File: lua/neo-tree/ui/renderer.lua [hunk 2]
```diff
@@ -938,7 +944,7 @@ local get_buffer = function(bufname, state)
     vim.api.nvim_buf_set_option(bufnr, "filetype", "neo-tree")
     vim.api.nvim_buf_set_option(bufnr, "modifiable", false)
     vim.api.nvim_buf_set_option(bufnr, "undolevels", -1)
-    autocmd.buf.define(bufnr, "WinLeave", function()
+    autocmd.buf.define(bufnr, "BufDelete", function()
```
Comment (cseickel) [inline lua/neo-tree/ui/renderer.lua:947]: Why was this changed? Does it still work if the buffer is not being displayed in a window?

File: lua/neo-tree/ui/renderer.lua (hunk 1 above)
Comment (pysan3) [reply lua/neo-tree/ui/renderer.lua:527]: No, we need to make `topline` or `lnum` falsy to skip the first line in save function.
Due to how the code is structured, the oldest save is most respected now. As I've written in the comment, when cursor position is overwritten, we should forget about this so that a later call to save can actually save its value. Otherwise, if the position is still saved, then the next call will not overwrite since oldest is greatest.

Additionally, I thought these `state.position.*` variables as private properties, and didn't like the idea to modify them from outside the scope.
That's why I made a public facing method to change these internal properties. Just like getters and setters in other languages.

File: lua/neo-tree/ui/renderer.lua (hunk 2 above)
Comment (pysan3) [reply lua/neo-tree/ui/renderer.lua:947]: When the buffer is not displayed in a window, users cannot change the position of the cursor right?

And for places that changes the cursor programatically (like follow current file), those cursor movements are captured in the _correct_ place (i.e. in `render_tree` or `close` etc).

Fun fact, that WinLeave _is_ the direct cause of the bug @mehalter was talking about in the issue.

File: lua/neo-tree/ui/renderer.lua (hunk 1 above)
Comment (cseickel) [reply lua/neo-tree/ui/renderer.lua:527]: I personally prefer the clear method to the `is.restorable` property, but it seems like we should not have both. Can we get rid of the `is.restorable` or is it handling another situation?

File: lua/neo-tree/ui/renderer.lua (hunk 2 above)
Comment (cseickel) [reply lua/neo-tree/ui/renderer.lua:947]: Makes sense.

File: lua/neo-tree/ui/renderer.lua (hunk 1 above)
Comment (pysan3) [reply lua/neo-tree/ui/renderer.lua:527]: As the comment says.

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/db178f4a49c19f8e4ed5a01dafa9d79e76f0081e/lua/neo-tree/ui/renderer.lua#L657-L659

This variable is more to decide whether we should restore the cursor position in the next iteration.
What `.clear()` does is more of a `is.savable + is.setable` (`save()` AND `set()` will do the thing next time), which stores if the next call to save should overwrite the previous saved `topline / lnum`.

But I don't like the name _savable_, like looking at the variable name, I don't quite get what the variable means at glance, or I have to read thru all the discussions that's held now to get the context.
Whereas `clear()` (to me) sounds like it forgets about the current cursor position and that's what it's literally doing.

Comment (cseickel) [reply lua/neo-tree/ui/renderer.lua:527]: I still don't see how setting `is.restorable = false` has a different outcome from using `clear()` in that situation and removing all references to `is.restorable`. If there is a subtle difference, then the system is too complicated.

Comment (pysan3) [reply lua/neo-tree/ui/renderer.lua:527]: Do you mean to depreciate `is.restorable` and use clear on every place? Then that's no problem, clear is a superset of that.

Comment (cseickel) [reply lua/neo-tree/ui/renderer.lua:527]: Yes, but I don't think there is no need to deprecate as this is all completely internal and shouldn't be used by end users. I searched and no external source I know of use it either.

Comment (pysan3) [reply lua/neo-tree/ui/renderer.lua:527]: Ok, I'll delete all occurrences of is.restorable and replace it with clear. 

Plz wait for a couple of days due to real life issues.

Comment (pysan3) [reply lua/neo-tree/ui/renderer.lua:527]: Done @cseickel ;)

**Review by cseickel — APPROVED** [review]
